python/lsst/ts/externalscripts/auxtel/latiss_base_align.py

```diff
@@ -0,0 +1,788 @@
+
+        self.log.info(
+            f"Intra/extra exposure ids: {self.intra_visit_id}/{self.extra_visit_id}"
+        )
+
+        self.angle = 90.0 - await self.atcs.get_bore_sight_angle()
+
+        self.log.info(f"Angle used in cwfs algorithm is {self.angle:0.2f}")
+
+        self.log.debug("Moving hexapod back to zero offset (in-focus) position")
+        # This is performed such that the telescope is left in the
+        # same position it was before running the script
+        await self.hexapod_offset(self.dz)
+
+    async def hexapod_offset(
+        self, offset: float, x: float = 0.0, y: float = 0.0
+    ) -> None:
+        """Applies z-offset to the hexapod to move between
+        intra/extra/in-focus positions.
+
+        Parameters
+        ----------
+        offset : `float`
+             Focus offset to the hexapod in mm.
+        x : `float`, optional
+            Offset the hexapod in the x-axis, in mm (default=0).
+        y : `float`, optional
+            Offset the hexapod in the y-axis, in mm (default=0).
+        """
+
+        offset = {
+            "m1": 0.0,
+            "m2": 0.0,
+            "x": x,
+            "y": y,
+            "z": offset,
+            "u": 0.0,
+            "v": 0.0,
+        }
+
+        self.atcs.rem.ataos.evt_detailedState.flush()
+        await self.atcs.rem.ataos.cmd_offset.set_start(
+            **offset, timeout=self.timeout_long
+        )
+        # Wait for ataos to go through a cycle, which will apply the offset if
+        # enough error has accumulated to pass the ataos thresholds.
+        # Success means we need to see ATAOS substate bit 3 (hexapod
+        # correction) or 4 (Focus correction), then 0 (IDLE).
+        # The bit(s) will flip when the correction is being determined,
+        # regardless  of if a hexapod actually had to move
+
+        start_time = time.time()
+        check_ss2 = False
+        while time.time() < start_time + self.timeout_long:
+            state = await self.atcs.rem.ataos.evt_detailedState.next(
+                flush=False, timeout=self.timeout_long
+            )
+            if (
+                bool(state.substate & (1 << 3))
+                or bool(state.substate & (1 << 4))
+                or check_ss2 is True
+            ):
+                check_ss2 = True
+                if state.substate == 0:
+                    break
+
+    def calculate_results(self) -> LatissAlignResults:
+        """Calculates hexapod and telescope offsets based on derotated
+        zernikes.
+
+        Returns
+        -------
+        results : `LatissAlignResults`
+            Results of the wavefront sensing.
+        """
+        rot_zern = np.matmul(
+            self.zern, self.matrix_rotation(self.angle + self.camera_rotation_angle)
+        )
+        hexapod_offset = np.matmul(rot_zern, self.matrix_sensitivity)
+        tel_offset = np.matmul(hexapod_offset, self.hexapod_offset_scale)
+
+        self.log.info(
```

Comment:
I really wish we could publish these results in a formatted way... like write an object to the LFA...
How does the MTAOS do it?
Can we be writing a file here?

Reply:
The MTAOS has events for these, but the results are also stored in the butler by the pipeline task. We can certainly publish them to the FLA tough.

I can make a ticket for that, don't want to make conflate too many changes in one go.